xrdp/xrdp_mm.c

```diff
@@ -1192,24 +1094,233 @@ dynamic_monitor_data(intptr_t id, int chan_id, char *data, int bytes)
+                          g_malloc(sizeof(struct dynamic_monitor_description), 1);
+
+    error = libxrdp_process_monitor_stream(s, &(dynamic_description->description), 1);
+    list_add_item(wm->mm->resize_queue, (tintptr)dynamic_description);
+    dynamic_description->start_time = g_time3();
+    advance_resize_state_machine(wm->mm->resize_state_machine, dynamic_description, WMRZ_QUEUED);
+    return 0;
+}
+
+const char *
+wm_login_state_to_str(enum wm_login_state login_state);
+
+/******************************************************************************/
+static int
+advance_error(int error, tbus resize_state_machine, struct dynamic_monitor_description *description)
+{
+    advance_resize_state_machine(resize_state_machine, description, WMRZ_ERROR);
+    return error;
+}
+
+/******************************************************************************/
+static int
+process_dynamic_monitor_description(struct xrdp_wm *wm, struct dynamic_monitor_description *description)
+{
+    int error = 0;
+    struct xrdp_mm *mm;
+    struct xrdp_mod *module;
+
+    LOG_DEVEL(LOG_LEVEL_TRACE, "process_dynamic_monitor_description:");
 
-    error = libxrdp_process_monitor_stream(s, description, 1);
-    if (error == 0)
+    if (wm == NULL)
     {
-        error = process_dynamic_monitor_description(wm, description);
+        return 1;
     }
-    g_free(description);
-    return error;
+    mm = wm->mm;
+    if (mm == NULL)
+    {
+        return 1;
+    }
+    module = mm->mod;
+    if (module == NULL)
+    {
+        return 1;
+    }
+
+    if (wm->login_state != WMLS_CLEANUP && wm->login_state != WMLS_INACTIVE)
+    {
+        LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: Not trying to resize while logging in! State was %s", wm_login_state_to_str(wm->login_state));
+        return 0;
+    }
+    if (wm->client_info->suppress_output == 1)
+    {
+        LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: Not allowing resize. Suppress output is active.");
+        return 0;
+    }
+    if (description->description.session_width <= 0 || description->description.session_height <= 0)
+    {
+        LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: Not allowing resize due to invalid dimensions (w: %d x h: %d)", description->description.session_width, description->description.session_height);
+        return 0;
+    }
+    if (description->state == WMRZ_QUEUED && description->description.session_width == wm->client_info->display_sizes.session_width && description->description.session_height == wm->client_info->display_sizes.session_height)
+    {
+        LOG(LOG_LEVEL_INFO, "process_dynamic_monitor_description: Not resizing. Already this size. (w: %d x h: %d)", description->description.session_width, description->description.session_height);
+        advance_resize_state_machine(mm->resize_state_machine, description, WMRZ_COMPLETE);
+        return 0;
+    }
+
+    switch (description->state)
```

Comment:
When I read the issue and commit description about a state machine to coordinate server screen resizes, I assumed that there would be a single resize state for the screen. Having the resize state stored in each resize request description is surprising since it means that the actual state of the screen (and supporting components) can be out of sync with the resize state enum in the resize description request.
Can you please explain why you went with sorting the resize state enum in the description  instead of with the screen that is being resized?

Reply:
My concern is that for the resize to be processed correctly we need to strictly process one resize request at a time to completion before starting the next resize request. With the current code it is possible to represent a resize state like the following: (and this example it would be a bug in the resize code)

```
xrdp_mm = {
    resize_queue = [
        dynamic_monitor_description = { display_resize_state = WMRZ_SERVER_VERSION_MESSAGE, ...},
        dynamic_monitor_description = { display_resize_state = WMRZ_QUEUED, ...},
        dynamic_monitor_description = { display_resize_state = WMRZ_ENCODER_DELETE, ...},
    ],
    ...
}
```

To avoid this situation the code has to be carefully written to ensure the invariants that only one dynamic_monitor_description has a state other than `WMRZ_QUEUED` at a time, and that 

The two ways that I deal with code that must be carefully written to avoid bugs is to encapsulate the critical path code that updates the state, and the other is to make invalid states unrepresentable. Here is how I would change the code to make the invalid state unrepresentable:
* move the `enum display_resize_state state` from the `dynamic_monitor_description` struct to the `xrdp_mm` struct
* add a `display_size_description display_description_resize_in_progress` to the `xrdp_mm` struct
* remove the explicit `WMRZ_QUEUED` state and replace its concept with the `display_size_description` just being in the resize_queue list

This would make the xrdp_mm struct look like:
```
xrdp_mm = {
    display_resize_state = WMRZ_SERVER_VERSION_MESSAGE,
    display_description_resize_in_progress = { /* display_size_description struct */ },

    resize_queue = [
        display_size_description = { /* display_size_description struct */ },
        display_size_description = { /* display_size_description struct */ },
    ],
    ...
}
```

Using this representation of the resize in progress it is much easier to enforce the invariant that there is only 1 resize in progress at a time, and the resize is processed to completion before the next resize is processed.

Reply:
I agree that the resize state is better stored in xrdp_mm, as apart from the resize being processed, every other item should be `WMRZ_QUEUED`, and so doesn't need an explicit state, which is another way of saying what @aquesnel has said above.

Using an explicit `display_description_resize_in_progress` is very clear, but you could also just use the head of the queue as the struct being worked on. There are advantages either way.  Using the head of the queue may save you a state, as you can use the queue length being > 0 as an indication that a resize is in progress.

Reply:
@matt335672 @aquesnel I spent a lot of time thinking about the right way to do this. I settled on this:
- Queue up only the `display_size_description` struct, as @aquesnel proposed.
- In xrdp_mm, now have `display_control_monitor_layout_data` which wraps `display_size_description` and contains the state and timestamps for logging. There is only one of these, and it is only ever populated if it is null. If it is not null, the state machine is processed as in the original version. This is a separate commit from the original commit so you can see how I've changed it.